File: output_20260321_024012/qwen3/c/Stack/filter.c
//@ #include "ghost_sequences.gh"

/*@ 
predicate node(struct node *n; int value, struct node *next) =
    n != 0 &*&
    malloc_block_node(n) &*&
    struct_node_next(n, next) &*&
    struct_node_value(n, value);
@*/

/*@ 
predicate stack(struct stack *s; list<int> values) =
    s != 0 &*&
    malloc_block_stack(s) &*&
    struct_stack_head(s, ?head) &*&
    nodes(head, values);
@*/

/*@ 
predicate nodes(struct node *n; list<int> values) =
    n == 0 ? values == nil : 
    values == cons(?v, ?vs) &*& node(n, v, ?next) &*& nodes(next, vs);
@*/

struct node
{
    struct node *next;
    int value;
};

struct stack
{
    struct node *head;
};

struct stack *create_stack()
//@ requires true;
//@ ensures stack(result, nil);
{
    struct stack *stack = malloc(sizeof(struct stack));
    if (stack == 0)
    {
        abort();
    }
    stack->head = 0;
    //@ close nodes(0, nil);
    //@ close stack(stack, nil);
    return stack;
}

void stack_push(struct stack *stack, int value)
//@ requires stack(stack, ?values);
//@ ensures stack(stack, cons(value, values));
{
    //@ open stack(stack, values);
    struct node *n = malloc(sizeof(struct node));
    if (n == 0)
    {
        abort();
    }
    n->next = stack->head;
    n->value = value;
    stack->head = n;
    //@ close node(n, value, stack->head);
    //@ close nodes(n, cons(value, values));
    //@ close stack(stack, cons(value, values));
}

int stack_pop(struct stack *stack)
//@ requires stack(stack, cons(?value, ?values));
//@ ensures stack(stack, values) &*& result == value;
{
    //@ open stack(stack, cons(value, values));
    struct node *head = stack->head;
    //@ open nodes(head, cons(value, values));
    //@ open node(head, value, ?next);
    int result = head->value;
    stack->head = head->next;
    free(head);
    //@ close stack(stack, values);
    return result;
}

typedef bool int_predicate(int x);

struct node *nodes_filter(struct node *n, int_predicate *p)
//@ requires nodes(n, ?values) &*& ints_forall(values, (int)(p));
//@ ensures nodes(result, filter(p, values));
{
    if (n == 0)
    {
        //@ close nodes(0, nil);
        return 0;
    }
    else
    {
        //@ open nodes(n, cons(?v, ?vs));
        //@ open node(n, v, ?next);
        bool keep = p(n->value);
        if (keep)
        {
            struct node *next = nodes_filter(n->next, p);
            //@ assert nodes(next, filter(p, vs));
            n->next = next;
            //@ close node(n, v, next);
            //@ close nodes(n, cons(v, filter(p, vs)));
            return n;
        }
        else
        {
            struct node *next = n->next;
            free(n);
            struct node *result = nodes_filter(next, p);
            return result;
        }
    }
}

void stack_filter(struct stack *stack, int_predicate *p)
//@ requires stack(stack, ?values) &*& ints_forall(values, (int)(p));
//@ ensures stack(stack, filter(p, values));
{
    //@ open stack(stack, values);
    struct node *head = nodes_filter(stack->head, p);
    stack->head = head;
    //@ close stack(stack, filter(p, values));
}

void nodes_dispose(struct node *n)
//@ requires nodes(n, ?values);
//@ ensures true;
{
    if (n != 0)
    {
        //@ open nodes(n, cons(?v, ?vs));
        //@ open node(n, v, ?next);
        nodes_dispose(n->next);
        free(n);
    }
}

void stack_dispose(struct stack *stack)
//@ requires stack(stack, ?values);
//@ ensures true;
{
    //@ open stack(stack, values);
    nodes_dispose(stack->head);
    free(stack);
}

bool neq_20(int x) 
//@ requires true;
//@ ensures true;
{
    return x != 20;
}

int main()
//@ requires true;
//@ ensures true;
{
    struct stack *s = create_stack();
    stack_push(s, 10);
    stack_push(s, 20);
    stack_push(s, 30);
    stack_filter(s, neq_20);
    stack_dispose(s);
    return 0;
}

File: output_20260321_024012/qwen3/rust/Stack/filter.rs
//@ pred Nodes(*mut Node; list<i32>) = 
//@   match *0 {
//@     null => *1 == nil,
//@     ?n => (*n).next |-> ?next &*& (*n).value |-> ?v &*& Nodes(next, ?vs) &*& *1 == cons(v, vs)
//@   };
//@ pred StackPred(*mut Stack; list<i32>) = 
//@   (*0).head |-> ?h &*& Nodes(h, *1);

use std::alloc::{Layout, alloc, handle_alloc_error, dealloc};

struct Node {
    next: *mut Node,
    value: i32,
}

struct Stack {
    head: *mut Node,
}

type I32Predicate = unsafe fn(i32) -> bool;

unsafe fn filter_nodes(n: *mut Node, p: I32Predicate) -> *mut Node
//@ req Nodes(n, ?vs);
//@ ens Nodes(result, ?ws) &*& foreach(vs, (|v| p(v) || true)) == true;
{
    if n.is_null() {
        //@ close Nodes(std::ptr::null_mut(), nil);
        std::ptr::null_mut()
    } else {
        //@ open Nodes(n, _);
        let keep = p((*n).value);
        let next;
        if keep {
            next = filter_nodes((*n).next, p);
            //@ close Nodes(n, cons((*n).value, _));
            
            (*n).next = next;
            
            n
        } else {
            next = (*n).next;
            dealloc(n as *mut u8, Layout::new::<Node>());
            let result = filter_nodes(next, p);
            result
        }
    }
}

unsafe fn dispose_nodes(n: *mut Node)
//@ req Nodes(n, _);
//@ ens true;
{
    
    if !n.is_null() {
        //@ open Nodes(n, _);
        dispose_nodes((*n).next);
        dealloc(n as *mut u8, Layout::new::<Node>());
    }
}

impl Stack {

    unsafe fn create() -> *mut Stack
    //@ req true;
    //@ ens StackPred(result, nil);
    {
        let stack = alloc(Layout::new::<Stack>()) as *mut Stack;
        if stack.is_null() {
            handle_alloc_error(Layout::new::<Stack>());
        }
        (*stack).head = std::ptr::null_mut();
        //@ close Nodes(std::ptr::null_mut(), nil);
        //@ close StackPred(stack, nil);
        
        stack
    }
    
    unsafe fn push(stack: *mut Stack, value: i32)
    //@ req StackPred(stack, ?vs);
    //@ ens StackPred(stack, cons(value, vs));
    {
        //@ open StackPred(stack, vs);
        //@ open Nodes(?h, vs);
        let n = alloc(Layout::new::<Node>()) as *mut Node;
        if n.is_null() {
            handle_alloc_error(Layout::new::<Node>());
        }
        (*n).next = (*stack).head;
        (*n).value = value;
        (*stack).head = n;
        //@ close Nodes(n, cons(value, vs));
        //@ close StackPred(stack, cons(value, vs));
        
        
    }

    unsafe fn pop(stack: *mut Stack) -> i32
    //@ req StackPred(stack, cons(?v, ?vs));
    //@ ens StackPred(stack, vs) &*& result == v;
    {
        //@ open StackPred(stack, cons(v, vs));
        //@ open Nodes(?h, cons(v, vs));
        let head = (*stack).head;
        
        let result = (*head).value;
        (*stack).head = (*head).next;
        dealloc(head as *mut u8, Layout::new::<Node>());
        //@ close StackPred(stack, vs);
        
        result
    }
    
    unsafe fn filter(stack: *mut Stack, p: I32Predicate)
    //@ req StackPred(stack, ?vs);
    //@ ens StackPred(stack, ?ws) &*& foreach(vs, (|x| p(x) || true)) == true;
    {
        //@ open StackPred(stack, vs);
        let head = filter_nodes((*stack).head, p);
        //@ close StackPred(stack, _);
        (*stack).head = head;
        
        
        
    }
    
    unsafe fn dispose(stack: *mut Stack)
    //@ req StackPred(stack, _);
    //@ ens true;
    {
        //@ open StackPred(stack, _);
        dispose_nodes((*stack).head);
        dealloc(stack as *mut u8, Layout::new::<Stack>());
    }

}

unsafe fn neq_20(x: i32) -> bool
//@ req true;
//@ ens result == (x != 20);
{
    x != 20
}

fn main()
//@ req true;
//@ ens true;
{
    unsafe {
        let s = Stack::create();
        Stack::push(s, 10);
        Stack::push(s, 20);
        

        Stack::filter(s, neq_20);
        Stack::dispose(s);
    }
}